### ISSUE TITLE
[RFC] `wash` command structure rework

### ISSUE THREAD

**Issue by brooksmtownsend** [wasmCloud/wash-archive#538]
As `wash` matures and gathers more features and functionality, the list of subcommands has gotten larger. We've also been using the tool for a while now, and we've got a good idea for what feels good about `wash` and what doesn't. This issue also serves as an RFC, and is primarily a proposal for a "redesigned" `wash` experience. There will be breaking changes here, but I think that these changes will be well worth it and serve to create a better experience for wash users in general.

At a high level, here are a few things that I feel create friction with the current CLI:
1. The most common commands are not at top-level subcommands
    - When I think of my _most commonly used_ commands, they are: build, ctl, new, inspect, up, and soon to be `app`. The fact that the `ctl` commands are often 3-4 levels deep nested is a shame since they are used so often
3. Some uncommon commands are top-level subcommands
    - completions, ctx, gen, lint, validate are all things that aren't commonly used except for one-time use or advanced cases
5. Verbage is inconsistent across the CLI, especially with far nested commands
    - `wash ctl link query` vs `wash ctl get claims` vs `wash ctl start actor`, these mix verb/noun order, subcommands to get information, etc. Confusing.

I think the best way to propose this general change in structure is to show you what I would love for the help text to look like, all comments welcome. 
```
~/github.com/ ➜ wash -h
_________________________________________________________________________________
                               _____ _                 _    _____ _          _ _
                              / ____| |               | |  / ____| |        | | |
 __      ____ _ ___ _ __ ___ | |    | | ___  _   _  __| | | (___ | |__   ___| | |
 \ \ /\ / / _` / __| '_ ` _ \| |    | |/ _ \| | | |/ _` |  \___ \| '_ \ / _ \ | |
  \ V  V / (_| \__ \ | | | | | |____| | (_) | |_| | (_| |  ____) | | | |  __/ | |
   \_/\_/ \__,_|___/_| |_| |_|\_____|_|\___/ \__,_|\__,_| |_____/|_| |_|\___|_|_|
_________________________________________________________________________________

Interact and manage wasmCloud applications, projects, and runtime environments


Usage: wash [OPTIONS] <COMMAND>

Applications:
  app          Manage declarative applications and deployments (wadm) (experimental)
  call         Invoke a wasmCloud actor
  link         Link an actor and a provider component with configuration
  start        Start an actor or a provider component on a running wasmCloud host
  stop         Stop an actor component, provider component, or host
  get          Retrieve information about running components or the lattice

Projects:
  build        Build (and sign) a wasmCloud actor, provider, or interface
  inspect      Inspect capability provider or actor module
  interface    Generate, lint, and validate smithy interfaces
  new          Create a new project from template
  push         Push an actor or provider component to an OCI or Bindle registry
  pull          Pull an actor or provider component from an OCI or Bindle registry
  par          Create, inspect, and modify capability provider archive files

Configuration:
  completions  Generate shell completions
  config       Manage wasmCloud host configuration contexts
  drain        Manage contents of local wasmCloud caches
  keys         Utilities for generating and managing keys

Runtime environments:
  up           Bootstrap a wasmCloud environment
  down         Tear down a wasmCloud environment launched with wash up

Options:
  -o, --output <OUTPUT>  Specify output format (text or json) [default: text]
  -h, --help             Print help
  -V, --version          Print version
```

I've organized the functionality into four sections, encapsulating the main responsibilities of `wash`, being (feedback welcome on wording):
1. Generate, Build, and Deploy wasmCloud projects
2. Interact with running wasmCloud applications
3. Manage settings, caches, downloads
4. Bootstrap wasmCloud runtime environments

The full command structure would be as follows. If anyone knows of a "CLI structure" that I can use instead of a bulleted list or YAML, happy to do so:
- app
  - list
  - get
  - history
  - del
  - put
  - deploy
  - undeploy
- call <actor> 
- link <actor> <provider>
- start <reference, fs/oci/bindle>
- stop <id>
- get
  - hosts
  - <host id for inventory>
  - claims
  - links
- build
- inspect <reference, fs/oci/bindle>
- interface
  - gen
  - lint
  - validate
- new
  - <name of project template to bootstrap, e.g. echo / messaging-provider>
- par
  - create
  - insert
- pull
- push
- completions
- config
  - TBD: Want to take inspiration from kubectl here
- drain
- keys
  - gen
  - get
  - list
- up
- down

One last note, I think `interface` and `par` will both be folded into `build` once we add support for building interfaces and providers.

Previous commands will hopefully get aliased so that we can remain compatible with as many existing scripts as possible, e.g. after this change `wash ctl start actor blahblah.azurecr.io/echo:0.1.0` would be equivalent to `wash start actor blahblah.azurecr.io/echo:0.1.0`, but we will output a warning that the previous structure will be removed in future minor (or perhaps a 1.0) releases.

**Comment by ricochet** [comment]
Most of the top-level commands are agnostic to type, except for those that interact with actors. Conveniently I also predict the following commands to be less frequently used as wadm adoption grows.

What do you think about moving the actor specific commands as a sub-command of the top-level type `actor`?
```bash
wash actor start
wash actor stop
wash actor call
```

**Comment by ricochet** [comment]
Actually, I've talked myself out of the above idea.

With the proposed design, the typical wash command is in the form of: `wash <action> [<group>] <item>`. I prefer this.

What I was struggling with is that apps which actors are grouped with follow `wash <group> <action>`, e.g. `wash app list` vs `wash start actor reference`. I changed my mind after asking myself what do we want users to primarily be doing with wash in wasmCloud environments? Building apps of course! App or apps should be a top-level command.

There isn't a hard and fast rule for designing CLI's. The way that you've thought about this by grouping commands is brilliant.

**Comment by ricochet** [comment]
The empty bullet under `get` feels like it could have all of the types listed, e.g. `actors`, `providers`, etc.

**Comment by ricochet** [comment]
The `reg` subcommand seems out of place. The action the user wants to do is "push up their build". The action they are thinking about is `push`, `release`, or maybe `deploy`. The name of the registry being used might be top of mind, but I've never seen `reg` as a command in other systems which makes it less discoverable and not representative of the action the user is attempting to take.

What do you think about moving these to top-level commands like start/stop?
 `wash push`
 `wash pull`

**Comment by andar1an** [comment]
Based on what you shared as `wash <action> [<group>] <item>` it would be something like `wash push reg item`? I like that if you can define the registry in group (an if group is optional there because you can set a default in config? Where are defaults set?)

**Comment by andar1an** [comment]
But then would the same logic hold for keys and par? etc?

**Comment by brooksmtownsend** [comment]
> What do you think about moving these to top-level commands like start/stop?
wash push
wash pull

I like it! In an ideal world it would be great to have something like `wash push github ./actor.wasm` but will have to consider that. For now I think the regular structure of `wash push <registry_url> <artifact_file>` and `wash pull <registry_url>` will work for now

Regarding your comment @stephenandary , we may want to consider a way to configure registries so that you can easily push to them without dragging out credentials each time. That could be an addition later that doesn't have to go in here, but you'll notice I nix'ed the `ctx` command in favor of `config` as a placeholder to better configure wash

**Comment by andar1an** [comment]
I saw that and loved it. Especially for enterprise use. Personally, I am using Cloudsmith for a private registry, and I think it is really important we make it easy for enterprise-type customers to work with private registries, whether cloudsmith, jfrog, etc.

**Comment by andar1an** [comment]
I don't know if this would be of interest, but I just learned about it and the first part of the README discusses command line flags: 
https://github.com/dtolnay/inventory

**Comment by ricochet** [comment]
I was pitching the next iteration to support `wash push <OCI ref>` without other arguments like the .wasm since in theory, wash just built it. A future pass with configuration would use the version of the actor and the registry configured in `wasmcloud.toml`, such that `wash push` will automatically fill in the registry url including organization and package, as well as set the tag to match the version of the actor. The inspiration is to feel similar to `cargo publish`.

- Passing in the file path feels like an override to me, e.g. `wash push ghcr.io/foo/bar:debug --artifact ./target/debug/actor.wasm`
- I'd prefer to not specify `github`, `docker` or any other registry as part of the command as that should be config.
- I strongly recommend all publish like commands have a `--dry-run` option

**Comment by brooksmtownsend** [comment]
@ricochet Wow I love that. Really sets us up to have something like
- wash build, builds and signs your thing!
- wash push, builds and signs and pushes your thing!
- wash start, builds and signs and pushes and starts your thing!

**Comment by ghost** [comment]
This looks great! I really like moving the `ctl` & `reg` commands to the top level. The biggest issue imo is simply the number of top level commands now - is that going to be too much for new users? But being top level might actually be better for new users than nested subcommands 🤔

Overall though LGTM!

**Comment by brooksmtownsend** [comment]
Closed with all the commits into v0.18!
